Fix bug where limit was happening before authorization check

TW80000 opened the pull request:
On calls to select_, AuthenticatedDAO decorates the sink with an AuthenticatedSink. The problem is that when the code eventually hits MDAO, it decorates the sink with a LimitedSink, which gets applied before the AuthenticatedSink instead of after.

For example, this means that if a user might have permission to see 12 objects in some DAO and they limit by 12, they might get less than 12 items back from the select, depending on the arbitrary order MDAO chooses to select items. It will pick the first 12 items and then do the auth checks, of which some might fail. The correct order is to first do the auth checks and then limit.

Commit:
Fix bug where limit was happening before authorization check

kgrgreer commented on src/foam/dao/AuthenticatedDAO.java:
```diff
@@ -99,27 +102,33 @@ public FObject find_(X x, Object id) {
   public Sink select_(X x, Sink sink, long skip, long limit, Comparator order, Predicate predicate) {
     // x.get("auth") != null added so that system user can act during booting when the auth service is not loaded yet.
     if ( authenticateRead_ && x.get("auth") != null ) {
```
If you move the authenticateRead_ && x.get()... check into augmentPredicate, it will make the code simpler and fix listen_.

TW80000 replied:
We don't want the `authenticateRead_` check to apply to `removeAll_`. I can move just the `x.get("auth") != null` check though.

TW80000 replied:
If I just remove the `x.get("auth") != null` check altogether, the server still runs and the tests still pass. Maybe whatever failure this was put in to fix has been fixed elsewhere since this was put in. Will just remove it instead.

TW80000 replied:
Resolved by 02d6fa3.

kgrgreer commented on src/foam/mlang/mlang.js:
```diff
@@ -3480,6 +3480,43 @@ return true;
+foam.CLASS({
+  package: 'foam.mlang.predicate',
+  name: 'HasPermission',
+  extends: 'foam.mlang.predicate.Binary',
```
Should this extend Binary? It isn't a normal Binary mlang, in that each argument isn't a Function, it just has two arguments. Then the args could be properly typed.

TW80000 replied:
That was something I wasn't sure about. Should it just extend `AbstractPredicate` then?

TW80000 replied:
Resolved by 65ef3a0.

kgrgreer commented:
I see.

On Fri, 12 Jul 2019 at 14:39, Troy Wolters <notifications@github.com> wrote:

> *@TW80000* commented on this pull request.
> ------------------------------
>
> In src/foam/dao/AuthenticatedDAO.java
> <https://github.com/foam-framework/foam2/pull/2395#discussion_r303103185>:
>
> > @@ -99,27 +102,33 @@ public FObject find_(X x, Object id) {
>    public Sink select_(X x, Sink sink, long skip, long limit, Comparator order, Predicate predicate) {
>      // x.get("auth") != null added so that system user can act during booting when the auth service is not loaded yet.
>      if ( authenticateRead_ && x.get("auth") != null ) {
>
> We don't want the authenticateRead_ check to apply to removeAll_. I can
> move just the x.get("auth") != null check though.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/foam-framework/foam2/pull/2395?email_source=notifications&email_token=ABY5PMYXFYVUFOHFIBK32V3P7DFWJA5CNFSM4ICORDX2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB6KOKJY#discussion_r303103185>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABY5PM2XTWN66NZY4XASL2TP7DFWJANCNFSM4ICORDXQ>
> .
>

kgrgreer commented:
yes

On Fri, 12 Jul 2019 at 14:43, Troy Wolters <notifications@github.com> wrote:

> *@TW80000* commented on this pull request.
> ------------------------------
>
> In src/foam/mlang/mlang.js
> <https://github.com/foam-framework/foam2/pull/2395#discussion_r303104468>:
>
> > @@ -3480,6 +3480,43 @@ return true;
>    ]
>  });
>
> +foam.CLASS({
> +  package: 'foam.mlang.predicate',
> +  name: 'HasPermission',
> +  extends: 'foam.mlang.predicate.Binary',
>
> That was something I wasn't sure about. Should it just extend
> AbstractPredicate then?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/foam-framework/foam2/pull/2395?email_source=notifications&email_token=ABY5PM7IFVTT7OXCVRB23IDP7DGD7A5CNFSM4ICORDX2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB6KOXIQ#discussion_r303104468>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABY5PM7LKNQ3K6QJ7M6YMODP7DGD7ANCNFSM4ICORDXQ>
> .
>

Commit:
Fix typo

Commit:
Give HAS_PERMISSION typed properties instead of extending Binary

As requested in a code review.

Commit:
Remove unnecessary null check

The server still runs and the tests still pass without this check.
Whatever it was protecting against must have been fixed.

adamvy commented on src/foam/dao/AuthenticatedDAO.java:
```diff
@@ -97,29 +100,34 @@ public FObject find_(X x, Object id) {
-    // x.get("auth") != null added so that system user can act during booting when the auth service is not loaded yet.
-    if ( authenticateRead_ && x.get("auth") != null ) {
-      super.select_(x, new AuthenticatedSink(x, createPermission("read"), sink), skip, limit, order, predicate);
+    if ( authenticateRead_ ) {
```
So you no longer have the `x.get("auth") != null` check anywhere right? I 100% support that because I think it could lead to security problems, however the comment implies that we needed it during boot?  Is that no longer the case?

TW80000 replied:
I think so. I removed it and started the server and I could sign in to the portal. Then I tried running the test suite and it passed. So based on that I'm assuming we don't need it anymore.
